pycloudlib/azure/cloud.py

```diff
@@ -0,0 +1,851 @@
+                    },
+                    "disable_password_authentication": True
+                }
+            },
+            "network_profile": {
+                "network_interfaces": [{
+                    "id": nic_id,
+                }]
+            },
+            "tags": {
+                "name": self.tag
+            }
+        }
+
+        if user_data:
+            # We need to encode the user_data into base64 before sending
+            # it to the virtual machine.
+            vm_parameters["os_profile"]["custom_data"] = base64.b64encode(
+                user_data.encode()).decode()
+
+        # We have two types of images, base images that are found on
+        # Azure marketplace and custom images based on other virtual
+        # machines. For marketplace images, we have an expected
+        # format: Canonical:UbuntuServer:19.10-DAILY. Here we have
+        # the published, offer and sku information. However, the
+        # snapshot images are just an id, similar to any other
+        # resource id that we find in Azure. Because of that
+        # difference, we need to treat those image types
+        # differently.
+        if image_id.startswith('Canonical'):
+            vm_parameters["storage_profile"]["image_reference"] = {
+                "publisher": 'Canonical',
+                "offer": util.get_offer_from_image_id(image_id),
+                "sku": util.get_sku_from_image_id(image_id),
+                "version": "latest"
+            }
+        else:
+            vm_parameters["storage_profile"]["image_reference"] = {
+                "id": image_id
+            }
+
+        return vm_parameters
+
+    def _create_virtual_machine(self, image_id, nic_id, user_data):
+        """Create a virtual machine.
+
+        This method provisions an Azure virtual machine for the image_id
+        provided by the user.
+
+        Args:
+            image_id: string, The image to be used when provisiong
+                      a virtual machine.
+            nic_id: string, The network interface to used for this
+                            virtual machine.
+            user_data: string, user data used by cloud-init when
+                       booting the virtual machine.
+        Returns:
+            The virtual machine created by Azure
+
+        """
+        vm_name = "{}-vm".format(self.tag)
+
+        self._log.debug('Creating Azure virtual machine')
+        vm_call = self.compute_client.virtual_machines.create_or_update(
+            self.resource_group.name,
+            vm_name,
+            self._create_vm_parameters(vm_name, image_id, nic_id, user_data)
+        )
+
+        return vm_call.result()
+
+    def delete_image(self, image_id):
+        """Delete an image from Azure.
+
+        Args:
+            image_id: string, The id of the image to be deleted
+        """
+        self._log.debug('Deleting Azure image: %s', image_id)
+        all_images = self.compute_client.images.list()
```

Comment:
It seems sort of costly to hit the `list()` method and walk through all images on these types of methods where we require an image_id param? I'm wondering why we wouldn't just call self.compute.client.images.get(image_name=image_id, resource_group_name=self.tag)

Maybe you are trying to be more flexible and searching all images in all resource groups so that they can be removed instead of a specific/known resource group?

I ask because this behavior differs from list_keys method for instance which is actually just looking within the self.resource_group.name

Reply:
My idea here was to extract the image_name from the Image object returned by azure, since in the API we need to the name of the image, instead of the id. But we can extract the name from the id, so I think it is better to use that approach.